### PR TITLE
15 implement transaction logic on program level

### PR DESCRIPTION
**Pull request by s-barabash**

### CODE REVIEW

**Review by OlehYakovenko — APPROVED** [review]

File: src/main/java/com/breskul/bibernate/action/Action.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package com.breskul.bibernate.action;
+
+public interface Action {
+
+  void execute();
+
+  int priority();
```
Comment (JoshukLite) [inline src/main/java/com/breskul/bibernate/action/Action.java]: Either add todo for documentation or the documentation itself.

File: src/main/java/com/breskul/bibernate/exception/DuplicateEntityException.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package com.breskul.bibernate.exception;
+
+public class DuplicateEntityException extends RuntimeException {
```
Comment (JoshukLite) [inline src/main/java/com/breskul/bibernate/exception/DuplicateEntityException.java:6]: Same as above

File: src/main/java/com/breskul/bibernate/transaction/Transaction.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+import com.breskul.bibernate.config.LoggerFactory;
+import com.breskul.bibernate.exception.BibernateException;
+import com.breskul.bibernate.persistence.Session;
+import java.sql.Connection;
+import java.sql.SQLException;
+import org.slf4j.Logger;
+
+//TODO: write tests
+public class Transaction {
```
Comment (JoshukLite) [inline src/main/java/com/breskul/bibernate/transaction/Transaction.java:18]: Docs

File: src/main/java/com/breskul/bibernate/transaction/Transaction.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+   * Commit the current resource transaction, writing any unflushed changes to the database.
+   *
+   * @throws IllegalStateException if <code>isActive()</code> is false
+   * @throws BibernateException    if the commit fails
+   */
+  public void commit() {
+    try {
+      validateIfSessionOpenAndTransactionActive();
+      connection.commit();
+      connection.setAutoCommit(true);
```
Comment (JoshukLite) [inline src/main/java/com/breskul/bibernate/transaction/Transaction.java]: Is it explicitly set without using the method below `setAutoCommitTrue`?

File: src/main/java/com/breskul/bibernate/transaction/Transaction.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+      throw new IllegalStateException("Cannot begin Transaction on closed session");
+    }
+    if (status != TransactionStatus.ACTIVE) {
+      throw new IllegalStateException("Cannot commit not active transaction");
+    }
+  }
+
+  private void validateIfCanRollback() throws SQLException {
+    if (!session.isOpen()) {
+      throw new IllegalStateException("Cannot begin Transaction on closed session");
```
Comment (JoshukLite) [inline src/main/java/com/breskul/bibernate/transaction/Transaction.java:106]: Is this message correct? Begin transaction during rollback?

File: src/main/java/com/breskul/bibernate/transaction/TransactionStatus.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
+package com.breskul.bibernate.transaction;
+
+public enum TransactionStatus {
```
Comment (JoshukLite) [inline src/main/java/com/breskul/bibernate/transaction/TransactionStatus.java:6]: Class docs

**Review by JoshukLite — CHANGES_REQUESTED** [review]

**Review by RomanPryshchepa — APPROVED** [review]

**Review by JoshukLite — APPROVED** [review]

**Pull request merged**